pkg/discovery/worker/garbage_collector.go

```diff
@@ -3,62 +3,108 @@ package worker
 import (
 	"context"
 	"fmt"
+	"strings"
 	"time"
 
 	"github.com/golang/glog"
 
 	api "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/kubernetes"
 
 	"github.com/turbonomic/kubeturbo/pkg/action/executor"
+	commonutil "github.com/turbonomic/kubeturbo/pkg/util"
 )
 
+var supportedGrandParents = []schema.GroupVersionResource{
+	schema.GroupVersionResource{
+		Group:    commonutil.K8sAPIDeploymentGV.Group,
+		Version:  commonutil.K8sAPIDeploymentGV.Version,
+		Resource: commonutil.DeploymentResName},
+	schema.GroupVersionResource{
+		Group:    commonutil.OpenShiftAPIDeploymentConfigGV.Group,
+		Version:  commonutil.OpenShiftAPIDeploymentConfigGV.Version,
+		Resource: commonutil.DeploymentConfigResName},
+}
+
+var supportedParents = []schema.GroupVersionResource{
+	schema.GroupVersionResource{
+		Group:    commonutil.K8sAPIReplicationControllerGV.Group,
+		Version:  commonutil.K8sAPIReplicationControllerGV.Version,
+		Resource: commonutil.ReplicationControllerResName},
+	schema.GroupVersionResource{
+		Group:    commonutil.K8sAPIReplicasetGV.Group,
+		Version:  commonutil.K8sAPIReplicasetGV.Version,
+		Resource: commonutil.ReplicaSetResName},
+}
+
+var gcListOpts = metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", executor.TurboGCLabelKey, executor.TurboGCLabelVal)}
+
 type GarbageCollector struct {
 	client                *kubernetes.Clientset
+	dynClient             dynamic.Interface
 	finishCollecting      chan bool
 	collectionIntervalSec int
 
 	// Necessary for test
 	podAge time.Duration
 }
 
-func NewGarbageCollector(client *kubernetes.Clientset, finishChan chan bool, collectionIntervalSec int, podAge time.Duration) *GarbageCollector {
+func NewGarbageCollector(client *kubernetes.Clientset, dynClient dynamic.Interface, finishChan chan bool, collectionIntervalSec int, podAge time.Duration) *GarbageCollector {
 	return &GarbageCollector{
 		podAge:                podAge,
 		client:                client,
+		dynClient:             dynClient,
 		finishCollecting:      finishChan,
 		collectionIntervalSec: collectionIntervalSec,
 	}
 }
 
-func (g *GarbageCollector) StartCleanup() {
-	glog.V(4).Info("Start leaked pods cleanup.")
-	// Also cleanup immediately at startup
-	g.cleanupLeakedClonePods()
-	g.cleanupLeakedWrongSchedulerPods()
+func (g *GarbageCollector) StartCleanup() error {
+	glog.V(4).Info("Start garbage cleanup.")
+
+	// We cleanup the controllers before cleaning up the pods to ensure
+	// that the pending (invalid controller) pods are deleted at the right time.
+	if err := g.CleanupControllers(); err != nil {
+		return nil
```

Comment:
Why return `nil` when there is an error?

Reply:
Thanks for catching this. Updated!